### ISSUE TITLE
Possible to generate an implementation that allows nulls?

### ISSUE THREAD

**Issue by renedewaele** [immutables/immutables#496]
I often run into the following issue. I want to define interfaces without `@Nullable` or optionals present for the end state of an object to make sure no required attribute is null. However, while building up this object (over the course of hours or sometimes days) I don't mind that some of its fields are null. I.e. I only want the not-null check to happen when the object is completely built.

I don't want to use a Builder for this because the Builder object is not meant to be queried or persisted and is mutable so inherently unsafe to use in events etc.

Basically, if specified by annotation it would be neat if the framework would generate an implementation of my interface that allows nulls and can be converted to the Immutable version (with null-checks) on demand. Note that the generated class may contain object attributes that also need to be converted to their Immutable version.

What do you think?

**Comment by renedewaele** [comment]
Another approach would be to add a style that doesn't force null-checks while the object is created but waits until a method (e.g. `isValid()`) is invoked, ideally also checking all attributes that have this style enabled.

**Comment by elucash** [comment]
@renedewaele Have you tried to use `@Value.Modifiable` for this? I somehow reluctant to gradually relax consistency and other guarantees in order to turn immutable value object into mutable, inconsistent, "anything can be null" kind of beans.
But modifiable companion objects are a little bit different, they are a sort of the builder or buffer (Like `ArrayList` for `ImmutableList`). Can have some attributes not initialized, but still non-null in its definition, then you can `toImmutable` and everything will be checked as one would expect.

**Comment by garakelian** [comment]
I agree completely with @renedewaele.  This is an essential feature for us, and I've just forked the repository to see if I can figure out how to disable this altogether (while maintaining immutability). Let me give you a use case, and try to convince you...

For our REST API, we use beans that look like this one for incoming requests:

`@ApiModel
	@Value.Immutable
	@JsonSerialize(as = ImmutableChangePasswordRequest.class)
	@JsonDeserialize(builder = ImmutableChangePasswordRequest.Builder.class)
	public static interface ChangePasswordRequest extends RestRequest {
		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "Old password", required = true, example = "0ldPassw0rd!", position = 100)
		@Nullable
		public String getOldPassword();

		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "New password", required = true, example = "N3wPassw0rd!", position = 110)
		@JsonProperty("newPassword")
		@Nullable
		public String getNewPassword();

		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "Confirm password", required = true, example = "N3wPassw0rd!", position = 120)
		@JsonProperty("confirmPassword")
		@Nullable
		public String getConfirmPassword();
	}`

As you can see, we've annotated the bean with NotNull, Size, and other constraints that are enforced by the Java Validation API.  This API is very sophisticated and can do much more than simple null checks.  

However, to use immutables, we've had to add a Nullable annotation as well (in logical conflict with the NotNull annotation), simply to get past the immutables generated code which absolutely insists that *IT* will be the final arbiter of what is required.  While I understand why this is desirable in many uses case, it's both frustrating and time consuming for us here.  And what's even worse, is if there is a primitive type.  In that case, we are *forced* to use Optional, something we don't want to do, because immutables will not allow you to apply @Nullable to a primitive field.

I enjoy the use of this library too much to go back now, so I plan to make the change for my own use.  I'd be happy to submit a pull request.  If you have any guidance on where you think it should be hooked in, let me know. Thanks.

**Comment by elucash** [comment]
@garakelian  I would love to do something to help your use case. But I'm trying to understand what kind of semantic you need. Is this a special kind of defaulting, where `null` will be used for objects, and `0`, `false` etc for primitives?
In order to implement this, we would try to switch the generator in a special mode, perhaps using style flag. Then, `isNullable()` would always return true for non-primitives. `isDefault` attribute model flag would be `true` to effectively disable "required" check. And the code, where usually `super.attr()` is called, can be changed to a default value `null`, `0`, `false`. See places where `DEFAULT_VALUE_` constants are used for annotation types.
Also, there should be understanding that you can construct such objects using a builder with incomplete/wrong data (until validated). This can be mitigated by using validation mix-in as shown [here](https://github.com/immutables/immutables/issues/26#issuecomment-242283196)

**Comment by garakelian** [comment]
@elucash Yes, I'm looking for a semantic where objects default to null, numeric primitives default to 0, booleans default to false, etc, exactly the way they would if I created a 'normal' bean and never call those setters. 

In my use case, I would like to use immutables to simplify the creation of "request objects", which capture the input from a HTTP REST request, and in the process greatly reduce the amount of boilerplate code that we have to write.  Immutable objects themselves are attractive for a number of reasons, mostly because the request received from a client should not be changed, in much the same way that HttpServletRequest has a getParameter(String) method, but not not have a setter (why would it, because you can only 'get' the value of a parameter received from a client, and not change the request).

The problem for me is the Immutables library insists on enforcing "mandatory" fields. I could work around this by tagging all of my fields with "Nullable" but there are two problems: (1) it's easy to forget to apply the Nullable annotation, resulting in unwanted validation, and (2) the library forbids you from using "Nullable" annotation on a primitives.  In many cases, I'm quite content with a default value of 0, or false, or null, as appropriate.

**Comment by garakelian** [comment]
@elucash Looking at the generated code, what I'm looking to do is turn off code generation related to initBits.

**Comment by elucash** [comment]
Yes, you can change those lines directly. Or make them implicitly nullable and default by changing the properties `ValueAttribute.isDefault` `ValueAttribute.isNullable()`. And then tweaking places where the actual default value is need to be inserted (clues like `DEFAULT_VALUE_ ` used for annotations)

Just out of curiosity, why not use struct-like classes? There are options to generate builders for plain objects: http://immutables.github.io/factory.html#pojo-constructors https://github.com/mkarneim/pojobuilder

```java
        @ApiModel
	public static interface ChangePasswordRequest extends RestRequest {
		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "Old password", required = true, example = "0ldPassw0rd!", position = 100)
		public String oldPassword;

		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "New password", required = true, example = "N3wPassw0rd!", position = 110)
		@JsonProperty("newPassword")
		public String newPassword;

		@NotNull
		@Size(min = 1, max = ValidationConstants.MAX_PASSWORD_LENGTH)
		@ApiModelProperty(value = "Confirm password", required = true, example = "N3wPassw0rd!", position = 120)
		@JsonProperty("confirmPassword")
		@Nullable
		public String confirmPassword;
	}
```

**Comment by garakelian** [comment]
@elucash, I considered the other library, but frankly, I was impressed by your documentation, support for Jackson, and the polite way you interacted with folks in the open issues (truth). I figured I start with your project :-).  Honestly, it worked flawlessly for me, except for this issue which I discovered while running some of my unit tests.  I figured that I'd try to contribute if I could do so within a short amount of time.

Jackson support is important to me because the Immutable beans I'm generating are used in a  Spring controller method that looks like this:
```
	public ResponseEntity<ChangePasswordResponse> handleChangePassword(
			@Valid @RequestBody final ChangePasswordRequest request) {
```

What templating language did you use for Immutables.generator.  It reminds me of Freemarker..

**Comment by twz123** [comment]
I just noticed that if a Immutable class is generated with `Value.Style.ValidationMethod.NONE`, then the with* methods on the immutable class are not null-safe unless explicitly annotated with `@Nullable`.

Is that intended behavior? Since there's no runtime check in place to ensure the non-nullness, we'll probably see NPEs again...

**Comment by elucash** [comment]
released in 2.4.0, thank you! we can probably remove null checks in case of NONE for withers and constructors, but this will be a separate bug/issue

**Comment by elucash** [comment]
@twz123 actually I've refined/fixed it to support constructors and `with` methods. Will be released in the next release
